ca/crl.go

```diff
@@ -0,0 +1,221 @@
+package ca
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"strings"
+	"time"
+
+	capb "github.com/letsencrypt/boulder/ca/proto"
+	"github.com/letsencrypt/boulder/core"
+	corepb "github.com/letsencrypt/boulder/core/proto"
+	"github.com/letsencrypt/boulder/issuance"
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+type crlImpl struct {
+	capb.UnimplementedCRLGeneratorServer
+	issuers   map[issuance.IssuerNameID]*issuance.Issuer
+	lifetime  time.Duration
+	maxLogLen int
+	log       blog.Logger
+}
+
+func NewCRLImpl(issuers []*issuance.Issuer, lifetime time.Duration, maxLogLen int, logger blog.Logger) (*crlImpl, error) {
+	issuersByNameID := make(map[issuance.IssuerNameID]*issuance.Issuer, len(issuers))
+	for _, issuer := range issuers {
+		issuersByNameID[issuer.Cert.NameID()] = issuer
+	}
+
+	if lifetime == 0 {
+		logger.Warningf("got zero for crl lifetime; setting to default 9 days")
+		lifetime = 9 * 24 * time.Hour
+	} else if lifetime >= 10*24*time.Hour {
+		return nil, fmt.Errorf("crl lifetime cannot be more than 10 days, got: %s", lifetime)
+	} else if lifetime <= 0*time.Hour {
+		return nil, fmt.Errorf("crl lifetime must be positive, got: %s", lifetime)
+	}
+
+	return &crlImpl{
+		issuers:   issuersByNameID,
+		lifetime:  lifetime,
+		maxLogLen: maxLogLen,
+		log:       logger,
+	}, nil
+}
+
+func (ci *crlImpl) GenerateCRL(stream capb.CRLGenerator_GenerateCRLServer) error {
+	var issuer *issuance.Issuer
+	var template *x509.RevocationList
+	var shard int64
+	rcs := make([]pkix.RevokedCertificate, 0)
+
+	for {
+		in, err := stream.Recv()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		switch payload := in.Payload.(type) {
+		case *capb.GenerateCRLRequest_Metadata:
+			if template != nil {
+				return errors.New("got more than one metadata message")
+			}
+
+			template, err = ci.metadataToTemplate(payload.Metadata)
+			if err != nil {
+				return err
+			}
+
+			var ok bool
+			issuer, ok = ci.issuers[issuance.IssuerNameID(payload.Metadata.IssuerNameID)]
+			if !ok {
+				return fmt.Errorf("got unrecognized IssuerNameID: %d", payload.Metadata.IssuerNameID)
+			}
+
```

Comment:
```suggestion
```

Reply:
My personal preference is to have a blank line after every `if err != nil {...}` or `if !ok {...}`. It serves to break the code into nice logical chunks: do as much setup as you like, then see if it worked, then break and move on to the next operation.